### PR TITLE
[@types/enzyme] adding `invoke` definition to support enzyme@3.10

### PR DESCRIPTION
**Pull request by rgibson518**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<
https://github.com/airbnb/enzyme/blob/b8289d02355882c57bf334ce001307807a674661/packages/enzyme/src/ShallowWrapper.js#L1305,
https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L837
>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

#### Misc
Just not sure how the process moves to support each version of Enzyme. I asked on their gitter, but they weren't really sure, either.  Do I need to bump the version in the `index.d.ts`?

#### Module Augmentation
I made this PR since I had already done a module augmentation to support `invoke`.  For those interested, here is the augmentation for the ShallowWrapper. I did not test this in my project using the (Common|React)Wrapper, however it should be trivial to make it work for them.
```typescript
import { Component } from 'react';

declare module 'enzyme' {
  interface ShallowWrapper<P = {}, S = {}, C = Component> {
    /**
     * Invokes a function prop.
     * @param invokePropName The function prop to call.
     * @param ...args The argments to the invokePropName function
     * @returns The value of the function.
     */
    invoke<K extends keyof P>(invokePropName: K): P[K];
  }
}
```

### CODE REVIEW

File: types/enzyme/index.d.ts
Comment (ljharb) [inline types/enzyme/index.d.ts]: `invoke` returns a function, which when again invoked, calls the prop function.

Thus, the type here should perhaps be (assuming P is the props):
```suggestion
    invoke<K extends keyof P>(invokePropName: K): P[K] extends Function
```
but ideally it'd also give you a type error when you passed a `K` such that `P[K]` isn't a function.

File: types/enzyme/enzyme-tests.tsx
Comment (jwbay) [inline types/enzyme/enzyme-tests.tsx]: Would like to this prop made optional so it's less invasive to the rest of the tests that don't care about invoke. If strictNullChecks is on, using ! is fine imo, or declaring a separate component/props for the invoke tests.

File: types/enzyme/enzyme-tests.tsx
Comment (jwbay) [inline types/enzyme/enzyme-tests.tsx]: Assignability can be a weak type-level assertion. Invoke could start returning any, for example, and all these tests would pass. Can you try `// $ExpectType number` instead for these? I'm not sure where the magic comment can be placed (above, same line, etc)

File: types/enzyme/index.d.ts
Comment (jwbay) [reply types/enzyme/index.d.ts]: The return type isn't currently a valid position for a constraint in TS. This seems to do what we want

```ts
type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends Function ? K : never }[keyof T];

interface Wrapper<P> {
    invoke<K extends keyof Pick<P, FunctionPropertyNames<P>>>(invokePropName: K): P[K]
}
```

**Review by jwbay — CHANGES_REQUESTED** [review]

File: types/enzyme/index.d.ts
Comment (rgibson518) [reply types/enzyme/index.d.ts]: Haha I was just working to update to use `FunctionPropertyNames`! Found that in the TS Handbook after a colleague at work mentioned it.  

This is what I had:
```typescript
/**
   * Invokes a function prop.
   * @param invokePropName The function prop to call.
   * @param ...args The argments to the invokePropName function
   * @returns The value of the function.
   */
  invoke<K extends NonNullable<{
      [K in keyof P]: P[K] extends () => void ? K : never
  }[keyof P]>>(invokePropName: K): P[K];
```

I inlined the `FunctionPropertyNames` because it seems that DT auto exports all declarations, so I wasn't sure if this was an export that should be owned by `@types/enzyme`. 

I was thinking of using `NonNullable` since the conditional allows for `undefined` and the inlined `FunctionPropertyNames` already "picks" those keys. I hadn't considered `Pick`. However, I do think `Pick` might make the intent more clear.  Thoughts on which to use?

File: types/enzyme/enzyme-tests.tsx
Comment (rgibson518) [reply types/enzyme/enzyme-tests.tsx]: After working with the aboves changes, I've started pulling out the "invoked" assignment tests. After thinking about it, these tests just prove that `functionProp` works as expect, not `invoke`.   I wrote these originally after looking at the enzyme docs for `invoke`. Being inspired by the `(...args)` in the docs, it felt necessary, but I'm feeling less so now.

File: types/enzyme/enzyme-tests.tsx
Comment (rgibson518) [reply types/enzyme/enzyme-tests.tsx]: I'll look into this, it felt bad updating so many tests that didn't rely on `functionProp`.

File: types/enzyme/enzyme-tests.tsx
Comment (ljharb) [reply types/enzyme/enzyme-tests.tsx]: is "$ExpectType" a TS thing, or a DT tests thing?

**Comment by rgibson518** [comment]
As a kind of tertiary improvement, thoughts on DRYing up the `invoke` tests into a `CommonWrapperHelper`? 

Example: 
```typescript
// outside of tests
const CommonWrapperHelper = {
    test_invoke: (wrapper: CommonWrapper<MyComponentProps, any, any>) => {
        // should only accept the names of props that are functions
        let is_matching_function: MyComponentProps['functionProp'] = wrapper.invoke('functionProp');
        // $ExpectError
        is_matching_function = wrapper.invoke('stringProp'); // should not accept non-function prop
        // $ExpectError
        const is_not_matching_function: () => string = wrapper.invoke('functionProp'); // should return matching function type
    },
};
// ShallowWrapper
function ShallowWrapperTest() {
    // ...other tests
    function test_invoke() {
        CommonWrapperHelper.test_invoke(shallowWrapper);
    }
};
// ReactWrapper
function ReactWrapperTest() {
    // ...other tests
    function test_invoke() {
        CommonWrapperHelper.test_invoke(reactWrapper);
    }
}
```

File: types/enzyme/enzyme-tests.tsx
Comment (rgibson518) [reply types/enzyme/enzyme-tests.tsx]: Seems to be part of [dtslint](https://github.com/microsoft/dtslint).

File: types/enzyme/index.d.ts
Comment (jwbay) [reply types/enzyme/index.d.ts]: 👍For inlining

**Comment by jwbay** [comment]
Do any other maintainers have any opinion here?

I'm a bit concerned about introducing abstractions in the tests that might make them harder to follow. There's already a lot going on in them. However, the current duplication is definitely and consistently annoying. I think I lean slightly toward your approach personally 👍

**Comment by rgibson518** [comment]
I have some more changes coming (as outlined in a few comments), please do not merge.

**Comment by rgibson518** [comment]
Still more coming... I think I need to make the `functionProps` less invasive.  Leaning toward extending the `MyComponentProps` with the `functionProps`, however I'm not 100% sure on this approach.

I'm also planning to prove that an undefined function prop should not be invokable.  At runtime, `invoke` will throw a TypeError if the prop is not a function, so I think it's reasonable to just "*prevent*" that with types in the first place.
```typescript
//example
interface Props {
  functionProp?(): void;
}
const MyComp: React.FC<Props> = ({ functionProp }) => <div></div>;
const wrapper = shallow(<MyComp/>);
// I believe invoke will accept `never` as argument since there are no props that satisfy it
// but I'm not very fluent in `never` nuances.
wrapper.invoke('functionProp') // $ExpectError
```

Or something like the above.

File: types/enzyme/enzyme-tests.tsx
Comment (jwbay) [inline types/enzyme/enzyme-tests.tsx]: Please revert indentation bumps like this as well

**Review by jwbay — CHANGES_REQUESTED** [review]
(Not sure why it marked this as approved)

**Comment by rgibson518** [comment]
> (Not sure why it marked this as approved)

Me either, lol..

File: types/enzyme/enzyme-tests.tsx
Comment (rgibson518) [reply types/enzyme/enzyme-tests.tsx]: Is there a formatter for DefinitelyTyped? Does the dtslinter have a --fix option or is there a prettier config? I was looking but didnt' see anything.

**Comment by rgibson518** [comment]
After some playing around with the function key selector portion of the invoke signature, I was able to get optional and required functions, with varying signatures (w/o using Function), while not getting optional non-function props.  I think this make the most sense. That said I think this is GTG unless of course there is more feedback!

File: types/enzyme/enzyme-tests.tsx
Comment (jwbay) [reply types/enzyme/enzyme-tests.tsx]: Not that I know of :(

**Review by jwbay — APPROVED** [review]

**Comment by cherealnice** [comment]
Any chance this gets merged soon? Currently blocking a project I'm working on, thanks!

**Comment by rgibson518** [comment]
@cherealnice Beyond merging, this will have to get published to @types/eznyme I think under the 3.10 tag.  However, I'm pretty sure if this PR is doesn't encompass everything needed to move to 3.10.  In the meantime you can use [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) if you're blocked. 

Add this to your project: 
```typescript
import { Component } from 'react';

declare module 'enzyme' {
  interface ShallowWrapper<P = {}, S = {}, C = Component> {
    /**
     * Invokes a function prop.
     * @param invokePropName The function prop to call.
     * @param ...args The argments to the invokePropName function
     * @returns The value of the function.
     */
    invoke<
      K extends NonNullable<
        { [K in keyof P]: P[K] extends ((...arg: any[]) => void) | undefined ? K : never }[keyof P]
      >
    >(
      invokePropName: K
    ): P[K];
  }
}
```

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]
@rgibson518 If you'd like this to be published as `3.10.0`, please bump the version at the top of `index.d.ts` to `3.10`. I'm not sure I follow your point about needing to add something to your code if you consume the package.

**Comment by rgibson518** [comment]
> @rgibson518 If you'd like this to be published as `3.10.0`, please bump the version at the top of `index.d.ts` to `3.10`. I'm not sure I follow your point about needing to add something to your code if you consume the package.

@uniqueiniquity I bumped the version since this covers a function from enzyme@3.10, this PR is definitely not comprehensive to all the changes from enzyme@3.10, though.  Perhaps it's better to leave it as 3.9 until 3.10 is more comprehensively covered?

**Comment by uniqueiniquity** [comment]
> @uniqueiniquity I bumped the version since this covers a function from enzyme@3.10, this PR is definitely not comprehensive to all the changes from enzyme@3.10, though. Perhaps it's better to leave it as 3.9 until 3.10 is more comprehensively covered?

That's fine; I think 3.10 is a better choice here.

Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Review by uniqueiniquity — APPROVED** [review]

**Pull request merged**
